### PR TITLE
deps: relax numpy pin, declare missing runtime deps, drop dead PIL imports

### PR DESCRIPTION
**Pull request by Steve-Allison**
## Summary

This PR makes `isanlp_rst` installable in projects that use **numpy 2.x**, fixes several runtime-imported packages that are missing from the declared dependency list, and removes two dead `from PIL import Image` lines.

The patch is intentionally tight (3 files, +6/-4) and focused on packaging/dependency hygiene only — no parser logic, models, or output schema is touched.

## Problem

`v3.2.0`'s `pyproject.toml` pins `numpy==1.26.4` exactly, which blocks installation in any environment that has migrated to numpy 2.x. Inspection of every numpy call site in the codebase (21 occurrences) shows usage limited to APIs that are stable across numpy 1.26 and 2.x:

- `np.unique(...)`
- `np.array(..., dtype=object)`
- `np.mean(...)`

There is no use of removed numpy 1.x aliases (`np.float`, `np.int`, `np.bool`, `np.string_`). The strict pin appears to be defensive over-pinning rather than reflecting an actual incompatibility.

Separately, four packages are imported at module load via the predictor's import chain but are not declared as dependencies — so `pip install isanlp_rst` succeeds, but `from isanlp_rst.parser import Parser` raises `ModuleNotFoundError` on first use:

| Package | Where it's imported |
|---|---|
| `huggingface_hub` | `dmrst_parser/predictor.py`, `universal_parser/predictor.py` |
| `tqdm` | both predictors |
| `networkx` | `dmrst_parser/src/parser/parsing_net.py`, `universal_parser/src/parser/parsing_net.py` |
| `pillow` | `rstviewer/main.py` (transitively loaded via `isanlp_rst/__init__.py`) |

Finally, two files contain `from PIL import Image` but never reference `Image` anywhere in the file — verified with `grep "Image\\." `: zero call sites in either.

### `pyproject.toml` (+8/-2)
- `numpy==1.26.4` → `numpy>=1.26,<3.0`
- Added `huggingface_hub`, `tqdm`, `networkx`, `pillow` to `dependencies`
- Removed `asyncio` (stdlib module — listing it is a no-op for installers)

### Source (-2 lines)
Removed dead `from PIL import Image` from:
- `isanlp_rst/dmrst_parser/src/parser/parsing_net.py:5`
- `isanlp_rst/universal_parser/src/parser/parsing_net.py:5`

`pillow` remains a hard dep because `rstviewer/main.py` legitimately uses `from PIL import Image, ImageChops`.

## Verification

- `pip install -e .` succeeds against a fresh Python 3.12 venv with `numpy 2.2.6` — no resolver conflicts.
- `from isanlp_rst.parser import Parser` imports cleanly.
- End-to-end parse using `tchewik/isanlp_rst_v3` with `hf_model_version='rstdt'` on a 32-word business text returns a valid `DiscourseUnit` tree:
  - `relation='Elaboration'`, `nuclearity='NS'`, span `0..212`
  - Both children present
  - All expected fields populated: `id`, `left`, `right`, `relation`, `nuclearity`, `start`, `end`, `text`, `proba`, plus `entropy`

## Test plan
- [ ] CI / maintainer reproduces `pip install -e .` on a numpy 2.x env
- [ ] Maintainer confirms parse output is identical pre/post (no logic changed)
- [ ] Optional: regenerate the `tchewik/isanlp_rst_v3` quick-start notebook against numpy 2.x to confirm
- [ ] Optional: maintainer adds CI job for numpy 2.x to prevent regression

## Out of scope (deliberately)
- `pytorch-lightning` (currently a hard dep, only used by `*/multiple_runs.py` training launchers) — could move to an optional `[training]` extra, but that's a maintainer-policy call worth a separate discussion.
- `requires-python = ">=3.8"` — the repo's actual `torch`/`transformers` versions floor much higher in practice. Worth raising to `>=3.10` or `>=3.11`, but again a maintainer policy call.
- Optional `[viewer]` extra for `playwright` / `lxml` — would be cleaner but requires lazy-import patches across multiple modules. Out of scope here.

## Context

I came across these issues while integrating `isanlp_rst` into a downstream project ([Story_Analyser](https://github.com/Steve-Allison/Story_Analyser)) which uses numpy 2.x. Maintaining a fork against `v3.2.0` for this purpose, but happy to upstream the dep-management fixes since they should benefit anyone in the same situation. Full background and verification artefacts in [my fork's FORK_NOTES.md](https://github.com/Steve-Allison/isanlp_rst/blob/story-analyser-numpy2-repin/FORK_NOTES.md) (intentionally not included in this PR — that's project-specific documentation).

Happy to iterate or split into smaller PRs if preferred.

### CODE REVIEW

**Comment by Steve-Allison** [comment]
Pushed a follow-up commit (`028ef9f`) addressing two things found while integrating into a downstream project. **No need to re-review the original commit** — it's unchanged.

## What's new in `028ef9f`

### 1. `np.string_` → `np.bytes_` on the training path

I missed this in the original numpy-usage audit because `training_manager.py` isn't loaded for inference. `np.string_` was removed in numpy 2.0; in numpy 1.x it was an alias for `np.bytes_`, so the patch is behaviour-preserving.

- File: `isanlp_rst/dmrst_parser/src/parser/training_manager.py:36-37`
- The matching line in `universal_parser/src/parser/training_manager.py:36` was already commented out upstream.
- Without this fix, `import isanlp_rst.dmrst_parser.src.parser.training_manager` crashes on numpy 2.x — anyone retraining on the modern numpy stack hits an ImportError before the first batch.

### 2. New: `tree_to_dict` / `tree_from_dict` JSON serialiser

Two small functions in a new `isanlp_rst/utils/serialization.py`, re-exported from `isanlp_rst.utils`. Useful for caching parsed trees, transmitting them over IPC, or persisting to disk without holding the live `DiscourseUnit` object graph.

Recursive serialiser preserving `id`, `relation`, `nuclearity`, `start`, `end`, `text`, `proba`, plus children. Round-trip-stable. `tree_from_dict` falls back to returning the dict unchanged when `isanlp` isn't importable, so dict-shaped consumers still work.

```python
from isanlp_rst.utils import tree_to_dict, tree_from_dict
import json

result = parser("Some text. More text.")
serialised = tree_to_dict(result['rst'][0])
restored = tree_from_dict(json.loads(json.dumps(serialised)))
```

This is more of a quality-of-life addition than a bug fix — happy to drop it from this PR if you'd prefer it as a separate, focused PR. It's contained in two new files (`utils/__init__.py`, `utils/serialization.py`) so removal is easy.

## Updated verification

In the same isolated Python 3.12 + numpy 2.2.6 venv:
- `parser("...")` → unchanged: `relation='Elaboration'`, `nuclearity='NS'`
- `import isanlp_rst.dmrst_parser.src.parser.training_manager` now succeeds (would have crashed before this commit)
- `tree_to_dict` round-trip OK on the parsed tree

## Updated diff stat

- Original commit: 3 files, +6/-4
- Follow-up: +3 files, +124/-1
- Total PR: 6 files changed

**Comment by Steve-Allison** [comment]
Pushed a third commit (`dacf07d`) addressing three more findings from continued downstream integration. **All cross-applicable to upstream — no Story_Analyser-specific API surface.** Original commits unchanged; no need to re-review them.

## What's new in `dacf07d`

### 1. `torch.load(weights_only=True)` — silences security FutureWarning

PyTorch 2.4+ emits a `FutureWarning` on every call to `torch.load()` because the default uses pickle (an arbitrary-code-execution vector). PyTorch 2.6+ flips the default to `weights_only=True`. State dicts from huggingface_hub are tensor-only, so explicitly setting `weights_only=True` is behaviour-preserving and pre-empts the upcoming default change.

- `isanlp_rst/dmrst_parser/predictor.py:91`
- `isanlp_rst/universal_parser/predictor.py:321`

### 2. Removed deprecated `verbose=True` from `ReduceLROnPlateau`

PyTorch 2.x removed this argument. The scheduler's native logging hook covers the same use case.

- `isanlp_rst/dmrst_parser/src/parser/training_manager.py:111`
- `isanlp_rst/universal_parser/src/parser/training_manager.py:151`

### 3. `print()` → structured logging in corpus utilities

The corpus modules write progress messages to stdout via `print()`. This breaks downstream callers that consume parser output via subprocess (their stdout becomes polluted). Replaced with `logging` calls so output goes to stderr and respects host log configuration.

- `{universal,dmrst}_parser/src/corpus/data.py` — `logger.debug` for per-file messages, `logger.info` for summary
- `{universal,dmrst}_parser/src/corpus/utils_dis_thiago.py` — preserved the existing `verbose` flag but routed through `logging` (verbose=True → INFO, verbose=False → DEBUG)

The `verbose` argument keeps the same control surface; only the output destination changes.

## Updated PR diff stat

| Commit | Files | +/- |
|---|---|---|
| `c4fda2c` (deps) | 3 | +6 / -4 |
| `028ef9f` (np.string_, tree_to_dict) | 3 | +124 / -1 |
| `dacf07d` (CVE, verbose, logging) | 8 | +130 / -64 |
| **Total** | **14 unique files** | **+260 / -69** |

## Splitting if you'd prefer

The three follow-up commits cover three independent concerns. Happy to split this into separate PRs if you'd prefer — just let me know and I'll close this and re-open them in sequence.

## Story_Analyser fork status

The fork now also contains additional API surface that I'm intentionally **not** sending upstream (device abstraction with MPS support, HF Hub cache controls, parse-result cache hook, parse_batch, parse_segments, type stubs, 40 pytest tests). They're feature additions that should be reviewed on their own merits — happy to discuss any of them in separate PRs if interesting.

**Comment by tchewik** [comment]
Thanks for the PR! Great cleanup and a helpful JSON serialization feature.

Batch parsing is long planned, but I haven't had time for it or even the refactoring yet. I'd be happy to add this and more to the library, but it may take some time to review properly.
